merkletree.go

```diff
@@ -83,11 +83,17 @@ func (m *MerkleTree) Get(key string) (MerkleNode, []ProofNode) {
 
 func (m *MerkleTree) Set(key string, value []byte) {
 	index := computePrivateIndex(key)
+
+	// generate random salt per user
+	salt := make([]byte, crypto.MaxSaltSize)
+	_, _ = io.ReadFull(rand.Reader, salt)
```

Comment:
I'm not sure what could go wrong here, but I would still read and catch the error here (*or* check if all bytes were read).
And a matter of style, but this could be changed to:
```
salt := make([]byte, c)
_, err := rand.Read(b)
```
`Read` uses  `ReadFull` internally: https://golang.org/src/crypto/rand/rand.go?s=714:752#L11